identity-core/src/common/timestamp.rs

```diff
@@ -21,44 +18,45 @@ use crate::error::Result;
 #[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
 #[repr(transparent)]
 #[serde(try_from = "String", into = "String")]
-pub struct Timestamp(DateTime<Utc>);
+pub struct Timestamp(OffsetDateTime);
 
 impl Timestamp {
   /// Parses a `Timestamp` from the provided input string.
   pub fn parse(input: &str) -> Result<Self> {
-    let datetime: DateTime<Utc> = DateTime::parse_from_rfc3339(input)?.into();
-    let datetime: DateTime<Utc> = Self::truncate(datetime);
-
-    Ok(Self(datetime))
+    let offset_date_time = OffsetDateTime::parse(input, &Rfc3339).map_err(time::Error::from)?;
+    Ok(Timestamp(truncate(offset_date_time)))
   }
 
   /// Creates a new `Timestamp` with the current date and time.
   pub fn now_utc() -> Self {
-    Self(Self::truncate(Utc::now()))
-  }
-
-  /// Returns the `Timestamp` as a Unix timestamp.
-  #[allow(clippy::wrong_self_convention)]
-  pub fn to_unix(&self) -> i64 {
-    self.0.timestamp()
-  }
-
-  /// Creates a new `Timestamp` from the given Unix timestamp.
-  pub fn from_unix(seconds: i64) -> Self {
-    Self(DateTime::from_utc(NaiveDateTime::from_timestamp(seconds, 0), Utc))
+    Self(truncate(OffsetDateTime::now_utc()))
   }
 
   /// Returns the `Timestamp` as an RFC 3339 `String`.
   ///
   /// See: https://tools.ietf.org/html/rfc3339
-  #[allow(clippy::wrong_self_convention)]
   pub fn to_rfc3339(&self) -> String {
-    self.0.to_rfc3339_opts(SecondsFormat::Secs, true)
+    // expect is OK, because all constructors provide well defined RFC 3339 compatible OffSetDateTime
+    // if this is to be made fallible instead then we need to change some things in our interface such as the
+    // From<Timestamp> for String implementation.
+    self
+      .0
+      .format(&Rfc3339)
+      .expect("unexpected Timestamp incompatibility with RFC 3339")
   }
 
-  fn truncate(value: DateTime<Utc>) -> DateTime<Utc> {
-    // Safe to unwrap because 0 is a valid nanosecond
-    value.with_nanosecond(0).unwrap()
+  /// Returns the `Timestamp` as a Unix timestamp.
+  pub fn to_unix(&self) -> i64 {
+    self.0.unix_timestamp()
+  }
+
+  /// Creates a new `Timestamp` from the given Unix timestamp.
+  pub fn from_unix(seconds: i64) -> Result<Self> {
+    let offset_date_time = OffsetDateTime::from_unix_timestamp(seconds).map_err(time::error::Error::from)?;
+    if !(0..10_000).contains(&offset_date_time.year()) {
+      return Err(time::error::Error::Format(time::error::Format::InvalidComponent("invalid year")).into());
```

Comment:
Maybe add a comment that this check is to prevent runtime errors in `to_rfc3339`?

This also needs unit tests for posterity. `OffsetDateTime::from_unix_timestamp` will already error on any timestamp outside the range `[-377705116800, 253402300799]` (i.e. `["-9999-01-01 0:00:00.0 +00:00:00", "9999-12-31T23:59:59Z"]`). So test some values in that range.

```suggestion
    // Reject years outside of the range 0000AD - 9999AD per Rfc3339
    // upfront to prevent conversion errors in to_rfc3339().
    // https://datatracker.ietf.org/doc/html/rfc3339#section-1
    if !(0..10_000).contains(&offset_date_time.year()) {
      return Err(time::error::Error::Format(time::error::Format::InvalidComponent("invalid year")).into());
```